src/nvim/drawscreen.c

```diff
@@ -1833,6 +1833,12 @@ static void win_update(win_T *wp, DecorProviders *providers)
       did_update = DID_NONE;
     }
 
+    if (wp->w_redr_statuscol) {
+      wp->w_redr_statuscol = false;
+      changed_line_abv_curs_win(wp);
+      goto win_update_start;
```

Comment:
This goto has caused horrible bugs in the past. Is there any way we can avoid it?

Reply:
It might be ok, it just raised an eyebrow. It's hard to tell without looking a lot more closely. Safer to avoid if we can.

Reply:
To be clear; we are restarting the loop because evaluating the `'statuscolumn'` resulted in a wider status column. I'll (re-)add a comment for that.

Reply:
So again, what do we want to avoid? The `goto`, specifically? Or restarting the loop? I had something like this before, which works as well:

```diff
     if (wp->w_redr_statuscol) {
       wp->w_redr_statuscol = false;
+      row = 0;
+      idx = 0;
+      lnum = wp->w_topline;
       changed_line_abv_curs_win(wp);
-      goto win_update_start;
+      redraw_later(wp, UPD_NOT_VALID);
+      continue;
     }
```

Would it be preferred? I only later discovered that there was a `goto` `label` already so I thought I would use that. The setup code before the main loop shouldn't really have to be executed again, so perhaps this would be better.

Reply:
An example where this caused a bug was #20665 which fixed a regression where the `goto` is called after important window state (`b_mod_top` and `b_mod_bot`) is cleared. The fix was to move the `goto` to before any of the window state is processed.

It looks like this code here calls the `goto` even later than that bug did.

Reply:
I see, that does sound problematic(I'll see if I can reproduce the issue somehow). Though for this purpose `b_mod_top/bot` should be 0 anyways, we want to redraw the entire window.

I just want to force a recursive redraw, I don't care how it happens😋. Is there another recommended way to do so? There is already a recursive redraw block at the end of `update_screen()`  but this is dependent on `cpotions` which also doesn't seem ideal.

If not, I think the `goto` or loop reset should be fine, perhaps also resetting `b_mod_top/bot` but I think these should already be 0.

Reply:
Yes, it seems irrelevant because this block will force an entire window update(we have changed `wp->w_nrwidth_width`:
https://github.com/neovim/neovim/blob/21ff00c06d34a8880b506e48943daaa172338370/src/nvim/drawscreen.c#L1046-L1048